### ISSUE TITLE
[BUG] Error: Sizes of tensors must match except in dimension 1. Expected size 64 but got size 512 for tensor number 2 in the list.

### ISSUE THREAD

**Issue by mikob** [microsoft/DeepSpeed#5091]
**Describe the bug**
When running an inference request, 

**To Reproduce**
1. Mask and image size of 512
2. What packages are required and their versions
```
python = ">=3.10,<3.12"
click = ">=7,<8"
supabase = "^2.3.4"
distro-info = "1.0"
scikit-build = "^0.16.7"
pillow = "^10.2.0"
ninja = "^1.11.1"
diffusers = "^0.26.2"
# optimization of stable-diffusion?
triton = "^2.2.0"
transformers = "^4.37.2"
pynsq = "^0.9.1"
compel = "^2.0.2"
scipy = "^1.12.0"
torch = "^2.2.0"
xformers = "^0.0.24"
accelerate = "^0.26.1"
deepspeed = { git = "https://github.com/microsoft/DeepSpeed.git", branch = "lekurile/fix_sd_ci" }
```

4. How to run the script
```
    with torch.inference_mode():
        pipe = StableDiffusionInpaintPipeline.from_pretrained(
            "/home/mikob/tools/stable-diffusion-v1-5-inpainting",
            # "/app/stable-diffusion-inpainting",
            # 'runwayml/stable-diffusion-inpainting'
            revision="fp16",
            torch_dtype=torch.float16,
            safety_checker=None,
        )

        pipe = pipe.to("cuda")
        pipe.safety_checker = dummy_safety_checker
        compel = Compel(tokenizer=pipe.tokenizer, text_encoder=pipe.text_encoder)

        # memory savings
        # pipe.enable_sequential_cpu_offload()
        # memory savings
        # pipe.enable_attention_slicing(1)
        # doesn't seem to have an effect
        # torch.backends.cuda.matmul.allow_tf32 = True
        # pipe.enable_xformers_memory_efficient_attention()

        pipe = deepspeed.init_inference(
            model=getattr(pipe,"model", pipe),
            mp_size=torch.cuda.device_count(),        # Number of GPUs
            dtype=torch.half, # dtype of the weights (fp16)
            replace_with_kernel_inject=True, # replace the model with the kernel injector
            # enable_cuda_graph=True, # enable the CUDA graph
        )

```
5. ...

**Expected behavior**
My inferences run fine without deepspeed, expect them to run the same way with it.

**Screenshots**

**System info (please complete the following information):**
 - OS: Linux Mint 21.2
 - GPU count and types 3070ti laptop 8gb

**Docker context**
Are you using a specific docker image that you can share?

**Additional context**

**Comment by lekurile** [comment]
Hello @mikob,

Using the latest DeepSpeed (installed from source) and Diffusers (0.26.2) versions, I'm able to get `runwayml/stable-diffusion-inpainting` working.

Here's the script I used:
```Python
from diffusers import StableDiffusionInpaintPipeline
import torch
from PIL import Image
import deepspeed
import os

local_rank = int(os.getenv("LOCAL_RANK", "0"))
device = torch.device(f"cuda:{local_rank}")

pipe = StableDiffusionInpaintPipeline.from_pretrained(
    "runwayml/stable-diffusion-inpainting",
    revision="fp16",
    torch_dtype=torch.float16,
)
pipe = pipe.to(device)

prompt = "Face of a yellow cat, high resolution, sitting on a park bench"
image = Image.open("./image.png")
mask_image = Image.open("./mask_image.png")

#image and mask_image should be PIL images.
#The mask structure is white for inpainting and black for keeping as is
image = pipe(prompt=prompt, image=image, mask_image=mask_image).images[0]
image.save("./yellow_cat_on_park_bench.png")

pipe = deepspeed.init_inference(
    pipe,
    mp_size=1,
    dtype=torch.half,
    replace_with_kernel_inject=True,
    enable_cuda_graph=True,
    )

deepspeed_image = pipe(prompt=prompt, image=image, mask_image=mask_image).images[0]
image.save("./deepspeed_yellow_cat_on_park_bench.png")
```

Please let me know if this script works for you as well when using the latest version of DeepSpeed.

Thanks,
Lev

**Comment by mikob** [comment]
@lekurile yes your script works. Thank you that's very helpful. I've narrowed down the things that break it, that otherwise work fine when deepspeed isn't used.

1) `Error: 'bool' object is not iterable `
```
def dummy_safety_checker(images, **kwargs): 
    return images, False 
pipe.safety_checker = dummy_safety_checker
```
Can work around this with `pipe.safety_checker = None`

2) Can't get a manual seed working:
Error: `Error: mat1 and mat2 must have the same dtype, but got Float and Half`
 ```
        seed = generator.seed()
        generator = generator.manual_seed(seed)

        image_latents = torch.randn(
            (count, 4, size // 8, size // 8),
            generator = generator,
            device = DEVICE
        )
        images = pipe(
            prompt_embeds=conditioning, 
        negative_prompt_embeds=negative_conditioning,
        width=size,
        height=size,
        latents=image_latents,
        ...
```

3) Issue with accepting image with 4 channels (rgba)
`Error: Sizes of tensors must match except in dimension 1. Expected size 64 but got size 512 for tensor number 2 in the list.`

Can be worked around this way
```
    photo_resized = photo.resize((size, size)).convert('RGB')
```

**Comment by lekurile** [comment]
@mikob,

If I understand correctly from the message above, issues 1 and 3 are resolved, right?

For issue 2, what seems to be happening is the data types aren't aligned somewhere. Can you ensure that the HF pipeline, DeepSpeed, and any other initialized data all conform to the same data format (e.g. Float, Half)?

Also, please feel free to share the updated code with the fixes in it's entirety if that helps.

Thanks,
Lev

**Comment by mikob** [comment]
@lekurile issue 1 is resolved with the workaround (shown above), but it shows that there is some incompatiblity with the API there so just wanted you guys to be aware. Because the same code works when I don't use deepspeed. Revisiting issue 3 actually I think had to do with inference_mode(), but likewise the workaround is there above. Issue 2 my only workaround is not using a manual seed. But anyway, the same code should work because it works without deepspeed.
